File: esim-frontend/test/workspaceSchematic.py
#
#  OpenESIM (Electrical Engineering Simulator)
#  Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>
#
#  This project is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public License(GPL)
#  as published by the Free Software Foundation; either version 2.1
#  of the License, or (at your option) any later version.
#
#  This project is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#

from PyQt5.QtWidgets import QAction, QActionGroup, QToolBar, QScrollArea
from PyQt5.QtGui import QIcon
from PyQt5.QtCore import Qt, QSize
from schematic.schematicWidget import SchematicWidget, ActionMode
from schematic.element import ElementDirection
from workspace import Workspace, Document

# --- Document ---
class DocumentSchematic(Document):
    def __init__(self, workspace, mainWindow, filename, symbolMode):
        #
        # Schematic Editor
        #
        schematicView = QScrollArea()
        self.schematicWidget = SchematicWidget(symbolMode, schematicView)
        schematicView.setWidget(self.schematicWidget)

        super().__init__(workspace, schematicView, filename)
        
        #
        # Actions - edit
        #
        undoStack = self.schematicWidget.getUndoStack()
        self.actionUndo = self._addWidget(undoStack.createUndoAction(mainWindow, "&Undo"))
        self.actionUndo.setIcon(QIcon(':/assets/undo.png'))
        self.actionUndo.setStatusTip('Undo.')
        
        self.actionRedo = self._addWidget(undoStack.createRedoAction(mainWindow, "&Redo"))
        self.actionRedo.setIcon(QIcon(':/assets/redo.png'))
        self.actionRedo.setStatusTip('Redo.')
        
        #
        # Menu - edit
        #
        mainWindow.menuEdit.addAction(self.actionUndo)
        mainWindow.menuEdit.addAction(self.actionRedo)
        
        #
        # Toolbar - edit
        #
        mainWindow.toolbarEdit.addAction(self.actionUndo)
        mainWindow.toolbarEdit.addAction(self.actionRedo)
        
        if False:
            for i in range(1):
                line = self.schematicWidget.createElement("line")
                line.setPos(120, 120, 200, 120)
                
                text = self.schematicWidget.createElement("text")
                text.setPos(128, 128)
                text.setText('Hello world 0')
                text.textStyle.set(QColor(0, 0, 0), 'Times New Roman', 8, False, False, False)
                text.setRotation(0)
                
                text = self.schematicWidget.createElement("text")
                text.setPos(128, 128)
                text.setText('Hello world 90')
                text.textStyle.set(QColor(0, 0, 0), 'Times New Roman', 8, False, False, False)
                text.setRotation(90)
                
                text = self.schematicWidget.createElement("text")
                text.setPos(128, 128)
                text.setText('Hello world 180')
                text.textStyle.set(QColor(0, 0, 0), 'Times New Roman', 8, False, False, False)
                text.setRotation(180)
                
                text = self.schematicWidget.createElement("text")
                text.setPos(128, 128)
                text.setText('Hello world 270')
                text.textStyle.set(QColor(0, 0, 0), 'Times New Roman', 8, False, False, False)
                text.setRotation(270)
                
                rect = self.schematicWidget.createElement("rect")
                rect.setPos(760, 300, 500, 500)
                
                ellipse = self.schematicWidget.createElement("ellipse")
                ellipse.setPos(400, 400, 600, 600)
                
                arc = self.schematicWidget.createElement("arc")
                arc.setAngle(30, -270*16)
                arc.setPos(200, 200, 300, 300)
                
                chord = self.schematicWidget.createElement("chord")
                chord.setAngle(0, -270*16)
                chord.setPos(300, 300, 400, 400)
                
                line = self.schematicWidget.createElement("line")
                line.setPos(120, 120, 100, 400)
                
                pin = self.schematicWidget.createElement("pin")
                pin.setPos(100, 400)
                pin.setDirection(ElementDirection.DIR_LEFT)
                pin = self.schematicWidget.createElement("pin")
                pin.setDirection(ElementDirection.DIR_RIGHT)
                pin.setPos(100, 400)
                
                pin = self.schematicWidget.createElement("pin")
                pin.setPos(100, 400)
                pin.setDirection(ElementDirection.DIR_TOP)
                
                pin = self.schematicWidget.createElement("pin")
                pin.setPos(100, 400)
                pin.setDirection(ElementDirection.DIR_BOTTOM)
                
            
            self.schematicWidget.update()
            
            fp = open("schematic1.sch", "w")
            self.schematicWidget.save(fp)
            
        if False:
            fp = open("schematic1.sch", "r")
            self.schematicWidget.open(fp)

    # @Override
    def open(self):
        #try:
        fp = open(self.filename, 'r')
        self.schematicWidget.open(fp)
        self.schematicWidget.update()
        fp.close()

        '''except:
            return False
        else:
            return True'''
            
    # @Override
    def save(self):
        #try:
        fp = open(self.filename, 'w')
        self.schematicWidget.save(fp)
        fp.close()
            
        '''except:
            return False
        else:
            return True'''
        
    # @Override
    def close(self):
        self.schematicWidget.close()

# --- Workspace ---
class WorkspaceSchematic(Workspace):
    def __init__(self, mainWindow, symbolMode):
        super().__init__(mainWindow)
        self.symbolMode = symbolMode
        #
        # Actions - mode
        #
        self.actionGroupModes = QActionGroup(mainWindow)
        
        self.actionSelection = self._addWidget(QAction(QIcon(':/assets/modeselection.png'), '&Enter selection mode.', mainWindow))
        self.actionSelection.setStatusTip('Enter selection mode.')
        self.actionSelection.setCheckable(True)
        self.actionSelection.triggered.connect(self.slotSelection)
        self.actionGroupModes.addAction(self.actionSelection)
        
        if self.symbolMode:
            self.actionInsertPin = self._addWidget(QAction(QIcon(':/assets/insertpin.png'), '&Insert a pin to schematic.', mainWindow))
            self.actionInsertPin.setStatusTip('Insert a pin to schematic.')
            self.actionInsertPin.setCheckable(True)
            self.actionInsertPin.triggered.connect(self.slotInsertPin)
            self.actionGroupModes.addAction(self.actionInsertPin)
        else:
            self.actionInsertDevice = self._addWidget(QAction(QIcon(':/assets/insertdevice.png'), '&Insert a device to schematic.', mainWindow))
            self.actionInsertDevice.setStatusTip('Insert a device to schematic.')
            self.actionInsertDevice.setCheckable(True)
            self.actionInsertDevice.triggered.connect(self.slotInsertDevice)
            self.actionGroupModes.addAction(self.actionInsertDevice)
            
        self.actionInsertLine = self._addWidget(QAction(QIcon(':/assets/insertline.png'), '&Insert a line to schematic.', mainWindow))
        self.actionInsertLine.setStatusTip('Insert a line to schematic.')
        self.actionInsertLine.setCheckable(True)
        self.actionInsertLine.triggered.connect(self.slotInsertLine)
        self.actionGroupModes.addAction(self.actionInsertLine)
        
        self.actionInsertRect = self._addWidget(QAction(QIcon(':/assets/insertrect.png'), '&Insert a rectangle to schematic.', mainWindow))
        self.actionInsertRect.setStatusTip('Insert a rectangle to schematic.')
        self.actionInsertRect.setCheckable(True)
        self.actionInsertRect.triggered.connect(self.slotInsertRect)
        self.actionGroupModes.addAction(self.actionInsertRect)
        
        
        self.actionInsertEllipse = self._addWidget(QAction(QIcon(':/assets/insertellipse.png'), '&Insert a ellipse to schematic.', mainWindow))
        self.actionInsertEllipse.setStatusTip('Insert a ellipse to schematic.')
        self.actionInsertEllipse.setCheckable(True)
        self.actionInsertEllipse.triggered.connect(self.slotInsertEllipse)
        self.actionGroupModes.addAction(self.actionInsertEllipse)
        
        self.actionInsertArc = self._addWidget(QAction(QIcon(':/assets/insertarc.png'), '&Insert an arc to schematic.', mainWindow))
        self.actionInsertArc.setStatusTip('Insert an arc to schematic.')
        self.actionInsertArc.setCheckable(True)
        self.actionInsertArc.triggered.connect(self.slotInsertArc)
        self.actionGroupModes.addAction(self.actionInsertArc)
        
        self.actionInsertChord = self._addWidget(QAction(QIcon(':/assets/insertchord.png'), '&Insert a chord to schematic.', mainWindow))
        self.actionInsertChord.setStatusTip('Insert a chord to schematic.')
        self.actionInsertChord.setCheckable(True)
        self.actionInsertChord.triggered.connect(self.slotInsertChord)
        self.actionGroupModes.addAction(self.actionInsertChord)
        
        self.actionInsertText = self._addWidget(QAction(QIcon(':/assets/inserttext.png'), '&Insert a text to schematic.', mainWindow))
        self.actionInsertText.setStatusTip('Insert a text to schematic.')
        self.actionInsertText.setCheckable(True)
        self.actionInsertText.triggered.connect(self.slotInsertText)
        self.actionGroupModes.addAction(self.actionInsertText)
        
        self.actionInsertOrigin = self._addWidget(QAction(QIcon(':/assets/insertorigin.png'), '&Insert an origin to schematic.', mainWindow))
        self.actionInsertOrigin.setStatusTip('Insert an origin to schematic.')
        self.actionInsertOrigin.setCheckable(True)
        self.actionInsertOrigin.triggered.connect(self.slotInsertOrigin)
        self.actionGroupModes.addAction(self.actionInsertOrigin)

        self.actionSelection.setChecked(True)
        
        #
        # Toolbar - mode 
        #
        self.toolbarMode = self._addWidget(QToolBar("Mode", mainWindow))
        self.toolbarMode.addAction(self.actionSelection)
        self.toolbarMode.addSeparator()
        if self.symbolMode:
            self.toolbarMode.addAction(self.actionInsertPin)
        else:
            self.toolbarMode.addAction(self.actionInsertDevice)
        self.toolbarMode.addSeparator()
        self.toolbarMode.addAction(self.actionInsertLine)
        self.toolbarMode.addAction(self.actionInsertRect)
        self.toolbarMode.addAction(self.actionInsertEllipse)
        self.toolbarMode.addAction(self.actionInsertArc)
        self.toolbarMode.addAction(self.actionInsertChord)
        self.toolbarMode.addAction(self.actionInsertText)
        self.toolbarMode.addSeparator()
        self.toolbarMode.addAction(self.actionInsertOrigin)
        self.toolbarMode.setIconSize(QSize(20, 20))
        mainWindow.addToolBar(Qt.LeftToolBarArea, self.toolbarMode)

    # @Override
    def switch(self, visible, currentDocument):
        super().switch(visible, currentDocument)

    # @Override
    def state(self):
        s = [self.actionSelection.isChecked(),
                self.actionInsertLine.isChecked(),
                self.actionInsertRect.isChecked(),
                self.actionInsertEllipse.isChecked(),
                self.actionInsertArc.isChecked(),
                self.actionInsertChord.isChecked(),
                self.actionInsertText.isChecked(),
                self.actionInsertOrigin.isChecked()]

        if self.symbolMode:
            s.append(self.actionInsertPin.isChecked())
        else:
            s.append(self.actionInsertDevice.isChecked())
        return s
        
    # @Override
    def setState(self, state):
        self.actionSelection.setChecked(state.pop(0))
        self.actionInsertLine.setChecked(state.pop(0))
        self.actionInsertRect.setChecked(state.pop(0))
        self.actionInsertEllipse.setChecked(state.pop(0))
        self.actionInsertArc.setChecked(state.pop(0))
        self.actionInsertChord.setChecked(state.pop(0))
        self.actionInsertText.setChecked(state.pop(0))
        self.actionInsertOrigin.setChecked(state.pop(0))
        
        if self.symbolMode:
            self.actionInsertPin.setChecked(state.pop(0))
        else:
            self.actionInsertDevice.setChecked(state.pop(0))
        
    # @Override
    def createDocument(self, mainWindow, filename):
        return DocumentSchematic(self, mainWindow, filename, self.symbolMode)
          
    #
    # Slot Processing
    #
    def slotSelection(self, event):
        pass
    def slotInsertPin(self, event):
        pass
    def slotInsertDevice(self, event):
        pass
    def slotInsertLine(self, event):
        pass
    def slotInsertRect(self, event):
        pass
    def slotInsertEllipse(self, event):
        pass
    def slotInsertArc(self, event):
        pass
    def slotInsertChord(self, event):
        pass
    def slotInsertText(self, event):
        pass
    def slotInsertOrigin(self, event):
        pass

File: esim-frontend/esim-ide/workspaceschematic.cc
/* ****************************************************************************
 * OpenESIM (Electrical Engineering Simulator)                                *
 * Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>                      *
 *                                                                            *
 * This project is free software; you can redistribute it and/or              *
 * modify it under the terms of the GNU Lesser General Public License(GPL)    *
 * as published by the Free Software Foundation; either version 2.1           *
 * of the License, or (at your option) any later version.                     *
 *                                                                            *
 * This project is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU          *
 * Lesser General Public License for more details.                            *
 * ****************************************************************************/

#include "workspaceschematic.h"
#include "mainwindow.h"
#include "schematicview.h"
#include "schematic.h"
#include "preview.h"
#include "errtrace.h"
#include <QAction>
#include <QActionGroup>
#include <QIcon>
#include <QUndoStack>
#include <QUndoView>
#include <QToolBar>
#include <QMenu>
#include <QMenuBar>
#include <QDockWidget>
#include <fstream>
#include <cstring>
#include <cassert>

namespace esim
{

////////////////////////////////////////////////////////////////////////////////
// Fixed widgets of schematic workspace                                       //
////////////////////////////////////////////////////////////////////////////////

WorkspaceSchematic::WorkspaceSchematic(MainWindow *mainWindow, bool symbolMode, const char *filename) :
    WorkspaceFixed(mainWindow, filename),
    m_symbolMode(symbolMode),
    m_scaleMax(5),
    m_scaleMin(0.5),
    m_scaleStep(0.2)
{
  /*
   * Schematic Editor
   */
  m_schematicView = this->addWidget(new schematic::SchematicView(m_symbolMode));
  connect(m_schematicView->view(), SIGNAL(modeChanged(int, const char *)), this, SLOT(slotModeChanged(int, const char *)));

  /*
   * Actions - edit
   */
  QUndoStack *undoStack = m_schematicView->undoStack();
  QAction *actionUndo = this->addAction(undoStack->createUndoAction(this, tr("&Undo")));
  actionUndo->setIcon(QIcon(":/assets/undo.png"));
  actionUndo->setStatusTip(tr("Undo."));

  QAction *actionRedo = this->addAction(undoStack->createRedoAction(this, tr("&Redo")));
  actionRedo->setIcon(QIcon(":/assets/redo.png"));
  actionRedo->setStatusTip(tr("Redo."));

  QAction *actionCopy = this->addAction(new QAction(QIcon(":/assets/copy.png"), tr("&Copy"), this));
  actionCopy->setIcon(QIcon(":/assets/copy.png"));
  actionCopy->setStatusTip(tr("Copy"));

  QAction *actionCut = this->addAction(new QAction(QIcon(":/assets/cut.png"), tr("&Cut"), this));
  actionCut->setIcon(QIcon(":/assets/cut.png"));
  actionCut->setStatusTip(tr("Cut"));

  QAction *actionPaste = this->addAction(new QAction(QIcon(":/assets/paste.png"), tr("&Paste"), this));
  actionPaste->setIcon(QIcon(":/assets/paste.png"));
  actionPaste->setStatusTip(tr("Paste"));

  /*
   * Actions View
   */
  QAction *actionZoomCenter = this->addAction(new QAction(QIcon(":/assets/zoomcenter.png"), tr("&Center at cursor"), this));
  actionZoomCenter->setStatusTip(tr("Center at cursor."));
  connect(actionZoomCenter, SIGNAL(triggered(bool)), this, SLOT(slotZoomCenter(bool)));

  QAction *actionZoomIn = this->addAction(new QAction(QIcon(":/assets/zoomin.png"), tr("&Zoom In"), this));
  actionZoomIn->setStatusTip(tr("Zoom In."));
  connect(actionZoomIn, SIGNAL(triggered(bool)), this, SLOT(slotZoomIn(bool)));

  QAction *actionZoomOut = this->addAction(new QAction(QIcon(":/assets/zoomout.png"), tr("&Zoom Out"), this));
  actionZoomOut->setStatusTip(tr("Zoom Out."));
  connect(actionZoomOut, SIGNAL(triggered(bool)), this, SLOT(slotZoomOut(bool)));

  QAction *actionZoomFit = this->addAction(new QAction(QIcon(":/assets/zoomfit.png"), tr("&Zoom Fit"), this));
  actionZoomFit->setStatusTip(tr("Zoom Out."));
  connect(actionZoomFit, SIGNAL(triggered(bool)), this, SLOT(slotZoomFit(bool)));

  /*
   * Actions - mode
   */
  QActionGroup *actionGroupModes = new QActionGroup(this);

  m_actionSelection = this->addAction(new QAction(QIcon(":/assets/modeselection.png"), tr("&Enter selection mode."), this));
  m_actionSelection->setStatusTip(tr("Enter selection mode."));
  m_actionSelection->setCheckable(true);
  connect(m_actionSelection, SIGNAL(triggered(bool)), this, SLOT(slotSelection(bool)));
  actionGroupModes->addAction(m_actionSelection);

  if (m_symbolMode)
    {
      m_actionInsertPin = this->addAction(new QAction(QIcon(":/assets/insertpin.png"), tr("&Insert a pin to schematic."), this));
      m_actionInsertPin->setStatusTip(tr("Insert a pin to schematic."));
      m_actionInsertPin->setCheckable(true);
      connect(m_actionInsertPin, SIGNAL(triggered(bool)), this, SLOT(slotInsertPin(bool)));
      actionGroupModes->addAction(m_actionInsertPin);
    }
  else
    {
      m_actionInsertDevice = this->addAction(new QAction(QIcon(":/assets/insertdevice.png"), tr("&Insert a device to schematic."), this));
      m_actionInsertDevice->setStatusTip(tr("Insert a device to schematic."));
      m_actionInsertDevice->setCheckable(true);
      connect(m_actionInsertDevice, SIGNAL(triggered(bool)), this, SLOT(slotInsertDevice(bool)));
      actionGroupModes->addAction(m_actionInsertDevice);
    }

  m_actionInsertLine = this->addAction(new QAction(QIcon(":/assets/insertline.png"), tr("&Insert a line to schematic."), this));
  m_actionInsertLine->setStatusTip(tr("Insert a line to schematic."));
  m_actionInsertLine->setCheckable(true);
  connect(m_actionInsertLine, SIGNAL(triggered(bool)), this, SLOT(slotInsertLine(bool)));
  actionGroupModes->addAction(m_actionInsertLine);

  m_actionInsertRect = this->addAction(new QAction(QIcon(":/assets/insertrect.png"), tr("&Insert a rectangle to schematic."), this));
  m_actionInsertRect->setStatusTip(tr("Insert a rectangle to schematic."));
  m_actionInsertRect->setCheckable(true);
  connect(m_actionInsertRect, SIGNAL(triggered(bool)), this, SLOT(slotInsertRect(bool)));
  actionGroupModes->addAction(m_actionInsertRect);

  m_actionInsertCircle = this->addAction(new QAction(QIcon(":/assets/insertellipse.png"), tr("&Insert a ellipse to schematic."), this));
  m_actionInsertCircle->setStatusTip(tr("Insert a ellipse to schematic."));
  m_actionInsertCircle->setCheckable(true);
  connect(m_actionInsertCircle, SIGNAL(triggered(bool)), this, SLOT(slotInsertCircle(bool)));
  actionGroupModes->addAction(m_actionInsertCircle);

  m_actionInsertArc = this->addAction(new QAction(QIcon(":/assets/insertarc.png"), tr("&Insert an arc to schematic."), this));
  m_actionInsertArc->setStatusTip(tr("Insert an arc to schematic."));
  m_actionInsertArc->setCheckable(true);
  connect(m_actionInsertArc, SIGNAL(triggered(bool)), this, SLOT(slotInsertArc(bool)));
  actionGroupModes->addAction(m_actionInsertArc);

  m_actionInsertChord = this->addAction(new QAction(QIcon(":/assets/insertchord.png"), tr("&Insert a chord to schematic."), this));
  m_actionInsertChord->setStatusTip(tr("Insert a chord to schematic."));
  m_actionInsertChord->setCheckable(true);
  connect(m_actionInsertChord, SIGNAL(triggered(bool)), this, SLOT(slotInsertChord(bool)));
  actionGroupModes->addAction(m_actionInsertChord);

  m_actionInsertText = this->addAction(new QAction(QIcon(":/assets/inserttext.png"), tr("&Insert a text to schematic."), this));
  m_actionInsertText->setStatusTip(tr("Insert a text to schematic."));
  m_actionInsertText->setCheckable(true);
  connect(m_actionInsertText, SIGNAL(triggered(bool)), this, SLOT(slotInsertText(bool)));
  actionGroupModes->addAction(m_actionInsertText);

  m_actionInsertOrigin = this->addAction(new QAction(QIcon(":/assets/insertorigin.png"), tr("&Insert an origin to schematic."), this));
  m_actionInsertOrigin->setStatusTip("Insert an origin to schematic.");
  m_actionInsertOrigin->setCheckable(true);
  connect(m_actionInsertOrigin, SIGNAL(triggered(bool)), this ,SLOT(slotInsertOrigin(bool)));
  actionGroupModes->addAction(m_actionInsertOrigin);

  m_actionSelection->setChecked(true);

  /*
   * Menu - edit
   */
  QMenu *m_menuEdit = mainWindow->menuEdit();
  m_menuEdit->addAction(actionUndo);
  m_menuEdit->addAction(actionRedo);
  m_menuEdit->addSeparator();
  m_menuEdit->addAction(actionCopy);
  m_menuEdit->addAction(actionCut);
  m_menuEdit->addAction(actionPaste);

  /*
   * Toolbar - edit
   */
  QToolBar *toolbarEdit = this->addWidget(new QToolBar(tr("Mode")));
  toolbarEdit->addAction(actionUndo);
  toolbarEdit->addAction(actionRedo);
  toolbarEdit->addSeparator();
  toolbarEdit->addAction(actionCopy);
  toolbarEdit->addAction(actionCut);
  toolbarEdit->addAction(actionPaste);
  toolbarEdit->setIconSize(MainWindow::toolIconSize());
  mainWindow->addToolBar(toolbarEdit);

  /*
   * Toolbar - view
   */
  QToolBar *toolbarView = this->addWidget(new QToolBar(tr("Mode")));
  toolbarView->addAction(actionZoomCenter);
  toolbarView->addAction(actionZoomIn);
  toolbarView->addAction(actionZoomOut);
  toolbarView->addAction(actionZoomFit);
  toolbarView->setIconSize(MainWindow::toolIconSize());
  mainWindow->addToolBar(toolbarView);

  /*
   * Toolbar - mode
   */
  QToolBar *toolbarMode = this->addWidget(new QToolBar(tr("Mode")));
  toolbarMode->addAction(m_actionSelection);
  toolbarMode->addSeparator();
  if (m_symbolMode)
    { toolbarMode->addAction(m_actionInsertPin); }
  else
    { toolbarMode->addAction(m_actionInsertDevice); }
  toolbarMode->addSeparator();
  toolbarMode->addAction(m_actionInsertLine);
  toolbarMode->addAction(m_actionInsertRect);
  toolbarMode->addAction(m_actionInsertCircle);
  toolbarMode->addAction(m_actionInsertArc);
  toolbarMode->addAction(m_actionInsertChord);
  toolbarMode->addAction(m_actionInsertText);
  toolbarMode->addSeparator();
  toolbarMode->addAction(m_actionInsertOrigin);
  toolbarMode->setIconSize(QSize(20, 20));
  mainWindow->addToolBar(Qt::LeftToolBarArea, toolbarMode);

  /*
   * Docker - Navigation
   */
  QDockWidget *navigation = this->addWidget(new QDockWidget);
  navigation->setWindowTitle(tr("Navigation"));
  schematic::PreviewSchematic *preview = new schematic::PreviewSchematic(m_schematicView->document(), navigation);
  navigation->setWidget(preview);
  mainWindow->addDockWidget(Qt::RightDockWidgetArea, navigation);

  /*
   * Docker - History
   */
  QDockWidget *history = this->addWidget(new QDockWidget);
  history->setWindowTitle(tr("History"));
  QUndoView *undoView = new QUndoView(undoStack, history);
  history->setWidget(undoView);
  mainWindow->addDockWidget(Qt::RightDockWidgetArea, history);
}

QWidget *WorkspaceSchematic::view() const
{
  return m_schematicView;
}

int WorkspaceSchematic::open()
{
  std::ifstream stream(this->filename().toLatin1().data());
  if (stream.is_open())
    {
      return m_schematicView->document()->open(stream);
    }
  return -RC_OPENFILE;
}

int WorkspaceSchematic::save()
{
  std::ofstream stream(this->filename().toLatin1().data());
  if (stream.is_open())
    {
      return m_schematicView->document()->save(stream);
    }
  return -RC_OPENFILE;
}

int WorkspaceSchematic::close()
{
  return 0;
}

bool WorkspaceSchematic::modified()
{
  return false;
}


void WorkspaceSchematic::slotSelection(bool checked)
{
  m_schematicView->view()->switchMode(schematic::EDIT_MODE_SELECT);
}
void WorkspaceSchematic::slotInsertPin(bool checked)
{

}
void WorkspaceSchematic::slotInsertDevice(bool checked)
{

}
void WorkspaceSchematic::slotInsertLine(bool checked)
{

}
void WorkspaceSchematic::slotInsertRect(bool checked)
{
  m_schematicView->view()->switchMode(schematic::EDIT_MODE_INSERT_ELEMENT, "rect");
}
void WorkspaceSchematic::slotInsertCircle(bool checked)
{
  m_schematicView->view()->switchMode(schematic::EDIT_MODE_INSERT_ELEMENT, "circle");
}
void WorkspaceSchematic::slotInsertArc(bool checked)
{
  m_schematicView->view()->switchMode(schematic::EDIT_MODE_INSERT_ELEMENT, "arc");
}
void WorkspaceSchematic::slotInsertChord(bool checked)
{

}
void WorkspaceSchematic::slotInsertText(bool checked)
{
  m_schematicView->view()->switchMode(schematic::EDIT_MODE_INSERT_ELEMENT, "text");
}
void WorkspaceSchematic::slotInsertOrigin(bool checked)
{
  m_schematicView->view()->switchMode(schematic::EDIT_MODE_INSERT_ELEMENT, "origin");
}

void WorkspaceSchematic::slotModeChanged(int mode, const char *classname)
{
  switch (mode)
  {
    case schematic::EDIT_MODE_SELECT:
      m_actionSelection->setChecked(true);
      break;
    default:
      return;
  }
}

void WorkspaceSchematic::slotZoomCenter(bool checked)
{

}

void WorkspaceSchematic::slotZoomIn(bool checked)
{
  if (m_schematicView->document()->getScale() < m_scaleMax)
    {
      m_schematicView->view()->scale(m_schematicView->document()->getScale() + m_scaleStep);
      m_schematicView->view()->update();
    }
}

void WorkspaceSchematic::slotZoomOut(bool checked)
{
  if (m_schematicView->document()->getScale() > m_scaleMin)
    {
      m_schematicView->view()->scale(m_schematicView->document()->getScale() - m_scaleStep);
      m_schematicView->view()->update();
    }
}

void WorkspaceSchematic::slotZoomFit(bool checked)
{
  m_schematicView->document()->setScale(1.0);
  m_schematicView->view()->update();
}

} // namespace esim
